[ca] dead code elimination for compile time

xmfan opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #141289
* #141153
* #141152

Although these nodes are eventually inlined away, they increase compile time, especially when initial CA graph capture treats all shapes as dynamic

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames @yf225

xmfan committed:
[ca] dead code elimination for compile time

[ghstack-poisoned]

jansel commented on torch/_dynamo/compiled_autograd.py:
```diff
@@ -305,18 +305,35 @@ def is_sym_node(self, node):
+
+        # Dynamo guards will error instead of creating aliasing guards unless we unpack them in the graph
+        unpack_nodes: Set[torch.fx.Node] = set()
+        i = len(self.graph_placeholders)
+        for node in self.fx_tracer.graph.nodes:
+            if i == 0:
+                break
+
+            unpack_nodes = unpack_nodes | set(node.users.keys())
+            i -= 1
```
Add some asserts about the node types here.  Use OrderedSet to be deterministic.

jansel commented on torch/_dynamo/compiled_autograd.py:
```diff
@@ -305,18 +305,35 @@ def is_sym_node(self, node):
+    def dce(self):
+        # Most of these removed nodes would have been removed during Dynamo and AOTDispatch
+        # Remove some of these nodes earlier to improve compilation speed
+        impure_targets = {
```
Put an OrderedSet in global scope.

xmfan committed:
Update on "[ca] dead code elimination for compile time"


Although these nodes are eventually inlined away, they increase compile time, especially when initial CA graph capture treats all shapes as dynamic

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames yf225

[ghstack-poisoned]

xmfan committed:
Update on "[ca] dead code elimination for compile time"


Although these nodes are eventually inlined away, they increase compile time, especially when initial CA graph capture treats all shapes as dynamic

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx chenyang78 kadeng chauhang amjames yf225

[ghstack-poisoned]

xmfan commented:
@pytorchbot merge
